raft/server.cc

```diff
@@ -509,11 +509,11 @@ future<entry_id> server_impl::add_entry_on_leader(command cmd, seastar::abort_so
 
     try {
         const log_entry& e = _fsm->add_entry(std::move(cmd));
-        memory_permit.release();
+        memory_permit.units.release();
```

Comment:
The semaphore units are used in the code because the code assumed sane implementation. If the implementation is broken it is much simpler to use the semaphore directly and release units explicitly in case of an exception.

Reply:
I've no objection if we ensure the semaphore is valid in all cases.

Reply:
Just check that we are still a leader before releasing.

Reply:
That won't work as is because the units are returned when destroyed (if not released).
We need another method like "break()" to gracefully detach the units from the semaphore after its destroyed.

Reply:
but what's so wrong with extending the semaphore lifetime as needed?
It's the cleanest and safest solution.

Reply:
> That won't work as is because the units are returned when destroyed (if not released). We need another method like "break()" to gracefully detach the units from the semaphore after its destroyed.

Do not use units. Units are used here as a RAII to release them in case of a local exception. If they API is not usable and not considered to be fixed I do not see the reason to use them. Use the semaphore directly.